Permissions rework

Thingus opened the pull request:
Closes #5091

### I have:

- [x] Formatted any Python files with [black](https://github.com/ambv/black)
- [x] Brought the branch up to date with master
- [x] Added any relevant Github labels
- [x] Added tests for any new additions
- [x] Added or updated any relevant documentation
- [x] Added an Architectural Decision Record (ADR), if appropriate
- [x] Added an [MPLv2 License Header](https://www.mozilla.org/en-US/MPL/headers/) if appropriate
- [x] Updated the [Changelog](https://github.com/Flowminder/FlowKit/blob/master/CHANGELOG.md) 

### Description

This PR reworks the way FlowAPI and FlowAuth handle query-scoped permissions.
At present, the permissions scopes are generated from a full walk of every set of query trees at every geographic level. This was causing memory issues with the growing number of queries, so this pull request implements a new system.

In this PR, `FlowAuth` consists of four entities; Servers, Users, Roles and Scopes.

```mermaid
erDiagram
    Servers ||--|{ Roles : contains
    Roles }|--|{ Scopes : "provides permissions for"
    Users }|--|{ Roles : "can work with"
    Servers ||--|{ Scopes : "provides"
```

Scopes are applied at the user level, giving permission to:
  * read results
  * run any queries at all
  * use specified queries, aggregations and location metrics to build their queries
  * (potentially) define the date range used to run their queries.
Scopes are automatically populated from a Server when it is registered or updated via a new FlowAPI endpoint
Roles are set by the FlowAuth admin, and consist of a set of Scopes. Roles are specific to Servers.
Users are assigned Roles by the FlowAuth admins.

The new auth flow runs as below:

```mermaid
sequenceDiagram
    participant Flowauth
    actor User
    participant FlowAPI
    note over User: Selects roles for token
    User ->>+ Flowauth: roles
    note over Flowauth: Checks user is permitted roles
    Flowauth ->>+ User: signed token w/ roles + scopes
    note over User: Builds query
    User ->>+ FlowAPI: query, token
    note over FlowAPI: Checks query params are covered by a single role
    FlowAPI ->>+ User: query results
```

This PR will remove Groups and Capabilities from FlowAuth, being replaced with Roles and Scopes. Tokens will also be removed as a database entitiy; instead, FlowAuth will treat a Token as a view of a combination of a User, a set of Roles and an expiry date.

greenape commented on flowapi/flowapi/permissions.py:
```diff
@@ -1,245 +1,64 @@
+    if isinstance(in_iter, dict):
+        in_iter = in_iter.values()
+    # Think there's a slightly better way of doing type introspection here
+    return all(type(item) not in [dict, list] for item in in_iter)
```
Could do `not any(isinstance(item, (dict, list)) for item in in_iter)` I guess? Not sure that's more satisfying really.

Thingus replied:
My other thought was `isinstance(Collection)` or similar, but something went wrong with that during dev.

greenape commented on flowapi/flowapi/permissions.py:
```diff
@@ -1,245 +1,64 @@
+            if node == key_of_interest:
+                # We cannot change the size of a dict mid-iterate, so instead we mark it for
+                # deletion post-iterate
+                root[node] = {}
```
Oo. I don't love the mutation. Must be a better way?

Thingus replied:
Yeah, I was racking my brains about this for a while - I'm going to circle back around to this once I've got the rest of the API in place, but there's gotta be a way of marking as part of the coroutine instead of the dict itself

greenape commented on flowapi/flowapi/permissions.py:
```diff
@@ -1,245 +1,64 @@
+def flatten_on_key(in_iter, key, _in_place=False):
+    if not _in_place:
+        in_iter = deepcopy(in_iter)
+    out = list(_flatten_on_key_inner(in_iter, key))
```
Why make it into a list here?

Thingus replied:
Logic being that this function is for returning a set of discrete 'things' for processing, a list made sense - I don't think you can make an assumption about what keys are relevant when this is applied. Probably should have been 'split_on_key'.

greenape replied:
Yeah, but I mean you could just glom the generators together, right? yield from a generator running `_clean_empties` over `_flatten_on_key_inner`?

Thingus replied:
Ah, point - I'll fix that up

greenape commented on flowkit_jwt_generator/flowkit_jwt_generator/jwt.py:
```diff
@@ -251,12 +252,14 @@ def generate_token(
         jti=str(uuid.uuid4()),
-        user_claims=compress_claims(claims) if compress else claims,
-        identity=username,
+        scopes=compress_claims(claims) if compress else claims,
```
Do we still need the compression?

Thingus replied:
Nah, I'm stripping it out as I go

greenape commented on flowauth/backend/flowauth/token_management.py:
```diff
@@ -153,12 +154,19 @@ def add_token(server):
+        )
+
+    allowed_roles = dict(filter(lambda x: x[1] is True, allowed_roles.items())).keys()
+    current_app.logger.debug(
```
I think for the tokens, good to have some debug logging, but should also log to a non-optional audit log that the token was generated?

Thingus replied:
:+1: Already in my local, long with what roles could have been responsible for the token

Thingus replied:
Ah, you mean to a specific `audit_logger` or similar; yes, I agree, I think that's a good idea.

jc-harrison commented on flowauth/backend/flowauth/models.py:
```diff
@@ -22,32 +22,36 @@
+scopes_in_role = db.Table(
+    "scopes_in_role",
+    db.Column("scope_id", db.Integer, db.ForeignKey("role.id"), primary_key=True),
+    db.Column("role_id", db.Integer, db.ForeignKey("scope.id"), primary_key=True),
```
Looks like the foreign keys are the wrong way around here.

jc-harrison commented on flowapi/flowapi/permissions.py:
```diff
@@ -267,83 +138,61 @@ def schema_to_scopes(schema: dict) -> Iterable[str]:
-        yield ".".join(paths)
+        logging.warning(
+            f"No aggregation unit options for {tl_query_name}; "
+            f"this should be fixed once PR 5278 is merged"
```
There will always be some top-level queries without an aggregation unit, because not all aggregates are spatial aggregates (currently `histogram_aggregate` is the only non-spatial aggregate, I think). I think we want something like `agg_units = ["nonspatial"]` in that situation.

jc-harrison commented on flowapi/flowapi/permissions.py:
```diff
@@ -267,83 +138,61 @@ def schema_to_scopes(schema: dict) -> Iterable[str]:
+            ["properties", "query_kind", "enum", 0],
+        )
+        if query_list == []:
+            return []
```
Some queries don't have any sub-queries, but we'll still need scopes for them (e.g. `<agg_unit>:<tl_query_name>:`)

Thingus replied:
Resolved: all queries now have the TL query as a sub_query (ex: `admin1:tl_query:tl_query`)

jc-harrison commented on flowauth/backend/flowauth/token_management.py:
```diff
@@ -153,28 +142,57 @@ def add_token(server):
-    current_app.logger.debug("New token request", allowed_claims=allowed_claims)
-    for claim in json["claims"]:
-        if claim not in allowed_claims:
-            raise Unauthorized(f"You do not have access to {claim} on {server.name}")
```
Need an equivalent of this in the new roles-based `add_token()`. I.e. `allowed_roles = current_user.allowed_roles(server)`, and check that all requested roles are in `allowed_roles`. Specifically, this needs to check:
1. Current user has access to `role`, and
2. `role` corresponds to the correct server (i.e. ensure we're not including scopes from a role on server B in a token valid for server A)

Thingus replied:
This is in-place with a test, `test_reject_claims_when_not_allowed`

jc-harrison replied:
Would be good to have a test for point 2 as well (i.e. reject if current user does have access to `role`, but `role` corresponds to the wrong server) - that's the kind of thing I can imagine we could easily break in future changes if we don't have a test for it.

Thingus replied:
Case 2 now tested in the same test

jc-harrison commented on flowauth/backend/flowauth/token_management.py:
```diff
@@ -153,28 +142,57 @@ def add_token(server):
+        role["name"]: db.session.execute(
+            select(Scope.name)
+            .join(Role.scopes)
+            .filter(Role.name == role["name"])
```
I'd like to have an additional filter here: `Role.server_id == server`.

jc-harrison commented on flowauth/frontend/src/TokenBuilder.jsx:
```diff
@@ -0,0 +1,185 @@
+    createToken(
+      name,
+      activeServer,
+      activeRoles.map(r => r.latest_token_expiry).reduce((prev,cur) => prev > cur ? prev : cur, new Date("1970")),
```
This should be finding the _earliest_ `latest_token_expiry` across the active roles, not the latest (i.e. the latest expiry allowed by _all_ active roles, not the latest expiry allowed by _any_ active role).

In fact, since we're no longer letting the user set the expiry in the UI, I think this bit of logic should be pushed into the backend (i.e. remove the 'expiry' parameter from the "Create token" API route, and instead calculate the expiry in `add_token()` in the backend).

Thingus replied:
In branch now

jc-harrison commented:
`UserObject.can_get_available_dates()` will need to be updated to check for the new 'get_available_dates' scope, rather than the old 'get_result:available_dates'.

Thingus commented on flowapi/flowapi/user_model.py:
```diff
@@ -24,26 +23,31 @@ class UserObject:
-                return True
+        granting_roles = []
+        for role, scopes in self.scopes.items():
+            if all(x in scopes for x in {*claims, *actions}):
```
I'm wondering if `get_result` should imply `can_poll`, or if we should just implement an explicit `poll` scope - it wouldn't be that much extra work to do. I can see one niche for it, if we have some kind of automatic provisioning or resource role that needs to keep an eye on what queries are running, but doesn't care what they are.

jc-harrison replied:
I think we want 'run' and 'get_result' to both imply 'poll', because in either of those situations you might need to be able to check the query state. I think the best solution is probably to just require the appropriate query scopes to poll a query, rather than requiring the query scopes _and_ one of ('run', 'get_result').

Thingus replied:
Yeah, that makes sense - I'll add it to the list

Thingus replied:
In branch + tested with `test_granular_poll_access`

Thingus commented:
Permissions system pull request + changelist

(created by referring to
git diff master...permissions_rework -- 'flowauth/backend'
...subbing in for folder path as appropriate)j

Overarching changes
------------------

- Permissions are now granted on 'role' - a 'role' is a collection
of 'scopes'
  - A role has an expiry date after which it cannot be used to grant scopes
  - A role has a maximum duration that tokens granted using it can last

- Each role is tied to one server.

- A 'user' has a set of 'roles'

- 'Groups' have been removed

- 'scopes' are now either simple or complex (still thinking on that)

- Simple scopes are one of the following strings: ('run', 
'get_available_dates', 'get_results') and represent a server-wide
capacity granted by that role.

- Complex scopes are a string: '[agg_unit]:[top-level scope]:[inner scope]'
granting access to a particular Flowmachine query
  - 'agg_unit'; An aggregation unit ('admin1 .. admin3, latlon, 'nonspatial')
('nonspatial' wwill be implemented on merge #5278#)
  - 'top_level query' is a flowmachine ExposedQuery
  - 'inner query' is a child query of ExposedQuery. Each ExposedQuery contains
itself as an inner query, (ex. 'admin1:daily_users:daily_users')


FlowAPI changes
--------------

- 'scopes' now nested inside 'roles' in tokens

/flowapi/flowapi/api_spec.py
- Default scopes are added to role here (53-55)

flowapi/config.py
- Default JWT_IDENTITY_CLAIM changed to 'sub' in line with JWT recs

flowkit/jwt_auth_callbacks.py
- identitiy key replaced with 'sub'

flowkit/permissions.py
- schema_to_scopes now works with the new scope system
- Rest has been rewritten to accomodate the new scope format

flowkit/user_model.py
- UserObject now takes a scopes dictionary
- has_access now works with the new roles, checking that one role
encompasses all requested scopes in the token


Flowauth backend changes
------------------------

- Roles have been added (models.py, roles.py)
- Groups have been removed (models.py, groups.py)
- longest_token_life replaced with longest_token_life_minutes
in all occurences

admin.py
- /tokens endpoint commented out (need to remove properly)

groups.py
- Removed

main.py
- Added an audit logger (check if I'm using this)
- Removed the groups blueprint
- Added '/roles' as the prefix for the interface to roles.py

models.py
- Added roles class + relationships (see diagram)
- Token renamed to TokenHistory (may not take this) 
- Edited the User class to use the Role latest and longest in validity check
- Changed Server from having Groups and Capabilities to Roles and Scopes
- Removed all GroupServer classes
- Changed make_demodata to use a limited set of Scopes and Users (may expand back to full?)
- Replaced 'owner' with 'user' and 'expires' with 'expiry'
- (Double check relationship cascade rules a: work and b: are appropriate)

roles.py
- Created
- Implements endpoints for roles editing:
  - listing all roles
  - listing roles for a user
  - getting a specific role
  - creating a new role
  - editing a role (should this merge scopes or replace them with the new ones?)
  - deleting a role
  - getting the members of a role
  - getting the scopes from a role
  - listing all roles for a user on a server
  - listing roles on a server

server.py
- Replaced 'capabilites' endpoint with 'roles' (replication from roles.py - where live?)
- Added endpoint that lists scopes on a server
- Added endpoint that updates scopes on a server
- Added endpoint for activating and deactivating scopes on a server (need to check if I
went through all the way with this or not)
- Added capabilites to edit_server to change the roles of a server
- (Does SQLAlchemy Cascade take care of deleting server roles?)

token_management.py
- rewrote list_my_servers()
- replaced 'expires' with 'expiry' in token args
- Implemented new token generation (Keep the currently commented-out API side checks?)

user_settings.py
- Added an endpoint that gets a set of roles for the logged-in user

users.py
- Stripped out references to groups
- Added 'roles' to edit_user endpoint


Flowauth frontend changes
--------

AdminMenu.jsx
-Groups replaced with Roles

Dashboard.jsx
- serverName removed for UserServer
- group_admin removed
- role_admin added

GroupDetails.jsx
- Removed

Picker.jsx
- Sanctioned abuse of Javascript functional programming to return the right value

RoleDetails.jsx
- Created
- View for creating and editing roles
- Used to define the following:
  - Role name
  - Server role is associated with
  - Users with a role (capability shared with User view)
  - Expiry date of the role
  - Scopes granted by a role
- Implemented in the functional React style

GroupList.jsx/RoleList.jsx
- Renamed GroupList to RoleList
- Changed the Lister from GroupDetails to RoleDetails

GroupMembersPicker.jsx/RoleMembersPicker.jsx
- Renamed GroupMemebersPicker to RoleMembersPicker
- Replaced getGroupMembers with getRoleMembers

RoleScopePicker.jsx
- Created
- Component for selecting scopes from a tree into a role
- Based on and similar to the existing ServerScopePicker
- Takes as properties a role_id and a server_id and an updateScopes callback
- Presents every scope available to server_id as an item in a CascadePicker
- Presents any role already available to role_id as already checked
- On update, passes new values to updateScopes
- There are extensive comments on the details of this, would encourage reviewer to read

ScopeDetails.jsx
- (Don't think I actually used this in the end, so can maybe remove?)

ServerAdminDetails.jsx
- Removed references to capabilites
- Added references to Scopes
- Updated scopeGraph signature 
- (This could do with some renaming of variables to make processes clearer)
- (Also lots of console.log() to remove)

TokenBuilder.jsx
- Created
- View used to make new tokens for a user by selecting their roles
- Only view reachable from a non-admin account
- Takes as properties a server and an onClick callback, usually used to return to prev. menu
- Used to define the following:
  - Token name (validated on update)
  - Roles to be included in token (why would a user _not_ click checkAll?)
- On successful submit, returns user to the server token management view, where the 
new token is visible for copy, view or download.

TokenDetails
- Rewritten to use roles + scopes
- Now deirves expiry from role that is furthest in the future (this should prob. be
other way around)

TokenList.jsx
- Added error log

TokensRolePicker
- Created
- Component for listing the roles available to build a token
- Takes as properties a list of roles, a checkAll function, a handleToggle function
and a 'checked' list of checked items

UserAdminDetails.jsx
- Removed groups
- Refactored createUser + editUser to use new API endpoint

UserRoleList.jsx
- Created
- Component used in TokenBuilder to provide a list of roles from a user
- Takes as properties:
  - a list of role objects
  - a checkAll function
  - a handleToggle callback
  - a 'checked' list containing the *indicies* of checked items in the list of role objects
  - a setRoleState callback (I think this is unused and can be removed)

UserRolePicker.jsx
- Created
- Component used in UserAdminDetails to pick roles for a given user
- Properties:
  - user_id - the ID of the user in question
  - updateRoles - a callback to call when the selected roles are changed  

UserServer.jsx
- Replaced TokenDetails with TokenBuilder view
- (Can prob delete TokenDetails view later, but want to use it to model tests on for now)

api.js
- Added 'roles' arg to editUser endpoint call
- Removed group-related functions
- Either removed capability-related functions or replaced 'capability' with 'scopes' or
'roles' as appropriate
- Replaced 'longest_token_life' with 'longest_token_life_minutes'
- Added endpoint calls for:
  - getting server scopes
  - getting all roles
  - getting specific roles
  - Creating a new role (returns the new role)
  - Deleting a role
  - getting roles for a user
  - getting roles for a user on a server
  - editing a role in it's entireity (editRole)
  - renaming a role (I don't think this is used and it can be removed)
  - editing role members
  - editing role scopes
    - All role edits call the same backend endpoint (PATCH '.../roles/) with a varying json payload
    - (I'm not sure that any apart from editRole are actually used, though)
 - Changed the createToken endpoint call to use roles instead of claims
 
util.js
- Almost full rewrite; most of the functions in here are scope-tree mangling of some form
- Rewrote jsonify to use the new scopes function
- Added docstrings to jsonify
- Added a scopesGraph function that turns a list of scopes into a nested tree
- Impleneted a function for finding the highest common roots of two trees, 
used in RoleScopePicker
- Implemented a function for rotating a list of scopes_with_roles to a list of scopes-from-
role, again used in the picker functions.

Still to do:
- Integrate James' PR
- Take James' comment on the poll query into consideration
- Check simple permissions work
- Check inside add_role for duplicate roles
- Check expiry rules for tokens 
- Unit tests for all new components
- Docstrings for new React components, paying special attention to component properties
- Check for and remove unneded bits from frontend
- Strip out many debug logs from frontend work

flowstef commented on flowauth/frontend/src/TokenRolesPicker.jsx:
```diff
@@ -0,0 +1,67 @@
+    <List>
+      {role.scopes.map((scope) =>
+      <ListItem>
+      <Typography varient = "body1" gutterTop>
```
typo: varient -> variant

jc-harrison commented on flowauth/frontend/src/util/util.js:
```diff
@@ -26,80 +27,144 @@ function zip() {
+    if (!(Object.keys(scopes_tree[admin_level]).includes(tl_query))){
+      scopes_tree[admin_level][tl_query] = {}
+    }
+    scopes_tree[admin_level][tl_query][inner_query] = true
```
Can this be implemented to handle ':'-delimited scopes with an arbitrary number of parts? If, in the future, we decided to add some 2-part scopes (or 4-part, etc.), it would be nice if FlowAuth was already able to handle these - then FlowAPI could include the new scopes in the API spec, and that API spec could be uploaded to an existing FlowAuth, without requiring a FlowAuth re-deployment.

jc-harrison commented on flowauth/frontend/src/util/util.js:
```diff
@@ -26,80 +27,144 @@ function zip() {
+      return {
+        label: this_label,
+        value: this_value,
+        enabled: this_value in enabled
```
Why does this need to set an 'enabled' key? Seems to me that the initial checkbox states are set by the `value` input to RightsCascade, so the jsonified graph just needs to capture the nesting structure.

Thingus replied:
A little bit of pre-engineering - I'm thinking that at some point, we may want to be able to turn scopes on or off from the server. Capturing `enabled` lets us filter here if it's turned on or off (iirc, everything is `enabled` by default atm, but this could change)

flowstef commented on flowapi/flowapi/user_model.py:
```diff
@@ -24,31 +23,39 @@ class UserObject:
         Name of the user
-    claims : dict
+    scopes : dict
         Dictionary giving a whitelist of the user's claims
```
We've never discussed this (I don't think) but do we as an organisation continue to use these terms (black-/whitelist)?

flowstef commented on flowapi/flowapi/user_model.py:
```diff
@@ -66,8 +73,8 @@ def can_run(self, *, query_json: dict) -> bool:
         """
-
-        return self.has_access(actions=["run"], query_json=query_json)
+        claims = await query_to_scopes(query_json)
```
A bit confusing that we use both "claims" and "scopes" here.

Thingus replied:
My thinking specifically for here:
`scopes` are what a user, via a role, *has* access to.
`claims` are what a user, via a query, is *trying* to access
There's ambiguity in that they both come in the same form (in this case, the output from query_to_scopes)

Having read the stackexchange link a few comments down, I realise that this is he exact opposite way around to what it should be - I'll switch them, and maybe refactor `query_to_scopes` to mean something else

jc-harrison replied:
In a JWT the "claims" are all the fields in the token body - in our case this includes the permission scopes, but also `"sub"`, `"exp"` etc.

I'd suggest using the term "scopes" everywhere here, whether you're referring to the scopes derived from a query spec or the scopes extracted from the JWT claims.

flowstef commented on flowapi/flowapi/user_model.py:
```diff
@@ -111,8 +118,8 @@ def can_poll(self, *, query_json: dict) -> bool:
         """
-
-        return self.has_access(actions=["run", "get_result"], query_json=query_json)
+        claims = await query_to_scopes(query_json)
```
Same as above: confusing terminology.

flowstef commented on flowapi/flowapi/user_model.py:
```diff
@@ -155,8 +161,8 @@ def can_get_results(self, *, query_json: dict) -> bool:
         """
-
-        return self.has_access(actions=["get_result"], query_json=query_json)
+        claims = await query_to_scopes(query_json)
```
As above, but this time with link :) https://security.stackexchange.com/questions/145776/what-are-openid-scopes-and-claims

Thingus replied:
Link copied to issue

flowstef commented on flowapi/flowapi/user_model.py:
```diff
@@ -210,8 +214,8 @@ def user_loader_callback(identity):
     )
-
-    claims = decompress_claims(get_jwt_claims())
+    # claims = decompress_claims(get_jwt_claims())
```
Why did you comment that out? If it's not needed I'd probably just remove it.

flowstef commented on flowauth/backend/flowauth/admin.py:
```diff
@@ -7,32 +7,33 @@
+#             for token in Token.query.all()
+#         ]
+#     )
+#
```
Any reason this is commented out as opposed to deleted?

flowstef commented on integration_tests/run_tests.sh:
```diff
@@ -22,7 +22,7 @@ if [ "$CI" != "true" ]; then
     (pushd .. &&  make down && make up && popd)
     echo "Waiting for flowdb to be ready"
-    docker exec flowdb bash -c 'i=0; until [ $i -ge 24 ] || (pg_isready -h 127.0.0.1 -p 5432); do let i=i+1; echo Waiting 10s; sleep 10; done'
+    docker exec flowdb bash -c 'i=0; until [ $i -ge 240 ] || (pg_isready -h 127.0.0.1 -p 5432); do let i=i+1; echo Waiting 10s; sleep 10; done'
```
Wow, quite a significant increase. How long does it usually take until the db is ready?

Thingus replied:
Think this was a bughunt change that escaped - I'll put it back

flowstef commented on flowkit_jwt_generator/flowkit_jwt_generator/jwt.py:
```diff
@@ -210,11 +210,12 @@ def squashed_scopes(scopes: List[str]) -> Iterable[str]:
     lifetime: datetime.timedelta,
     claims: List[str],
-    compress: bool = True,
+    compress: bool = False,
```
Why don't we compress anymore?

jc-harrison replied:
Ah, I can answer that one. We started compressing the token claims when the old permissions got so big that tokens no longer fitted in http headers. It wasn't ideal (meant that users could no longer directly check the permissions granted by a token just from the decoded token - they would also have to decompress the claims), but was a necessary workaround at the time. But with the new (smaller) permissions, that workaround is no longer necessary.

flowstef commented on flowauth/frontend/src/UserServer.jsx:
```diff
@@ -39,10 +39,10 @@ class UserServer extends React.Component {
           classes={this.props.classes}
           onClick={this.toggleEdit}
-          serverID={this.props.serverID}
+          activeServer={this.props.serverID}
```
Why rename only half of it?

Thingus replied:
The serverID prop is used elsewhere, so I didn't want to touch it.

flowstef commented on flowauth/frontend/src/ScopeDetails.jsx:
```diff
@@ -0,0 +1,28 @@
+
+  return (
+    <Fragment>
+      <ListItem varient="h6">{scope.name}
```
typo: variant

flowstef commented on flowauth/frontend/src/ScopeDetails.jsx:
```diff
@@ -0,0 +1,28 @@
+      <List>
+        {scope.roles.map((role) =>
+        <ListItem>
+        <Typography varient = "body1" gutterTop>
```
typo: variant

flowstef commented on flowauth/backend/tests/test_version_route.py:
```diff
@@ -5,9 +5,8 @@
 
 
-def test_version(client):
+def test_version(client, app):
```
not sure why you add `app` here but don't use it.

Thingus replied:
I think that importing the fixture also provides some setup

Thingus replied:
I took it out and it broke the test :sweat_smile: so I'll leave it in

flowstef commented on flowauth/backend/tests/test_token_admin.py:
```diff
@@ -7,15 +7,15 @@
+#     response, csrf_cookie = auth.login(uname, upass)
+#     response = client.get("/admin/tokens", headers={"X-CSRF-Token": csrf_cookie})
+#     assert 200 == response.status_code
+#     assert 1 == len(response.get_json())
```
Why is this commented out instead of deleted?

flowstef commented on flowauth/backend/tests/test_auth.py:
```diff
@@ -52,11 +53,12 @@ def test_is_logged_in(client):
-    with client:
-        auth.logout()
-        assert "user_id" not in session
+        with client:
```
From the python docs:

> With more than one item, the context managers are processed as if multiple [with](https://docs.python.org/3/reference/compound_stmts.html#with) statements were nested

https://docs.python.org/3/reference/compound_stmts.html#with

flowstef replied:
I think that would look nicer :)

Thingus replied:
Breaks the test, though :-(

flowstef commented on flowauth/backend/tests/test_access_reflects_server_limits.py:
```diff
@@ -2,17 +2,17 @@
 )
 
 
+@pytest.mark.skip(reason="No groups anymore; not sure if this is relevant")
```
I think if it's no longer relevant we should remove it instead of skipping.

greenape replied:
presumably still relevant but for user's instead of groups?

jc-harrison replied:
Still relevant for roles, I think.

flowstef commented on flowauth/backend/flowauth/user_settings.py:
```diff
@@ -225,6 +227,18 @@ def is_two_factor_active():
+    """
+    server = Server.query.filter(Server.id == server_id).first_or_404()
+    roles = Role.query.filter(Role.server == server).join(User.roles).all()
+    # NOTE: This includes other users inside the role, leading to enumeration. Is this a problem?
```
It doesn't sound clean but others may have a better idea.
Can you not filter the other users out?

flowstef commented on flowapi/tests/unit/test_permissions.py:
```diff
@@ -123,66 +105,161 @@ def test_valid_tree_walks(tree, expected):
-                "get_result&available_dates",
+                "unset:dummy:dummy",
+                "unset:dummy:nested_dummy",
+                "unset:dummy:nested_dummy_2",
```
Can I just say: well done! This is so much more readable!!

flowstef commented on flowauth/backend/tests/test_commands.py:
```diff
@@ -2,8 +2,12 @@
+    add_admin_command,
+    demo_data,
+    init_db_command,
+)  # , add_role_command
```
why comment out?

flowstef commented on flowauth/backend/tests/test_commands.py:
```diff
@@ -105,4 +109,17 @@ def test_demo_data(app):
+        assert len(Scope.query.all()) == 3
+
+
+# def test_add_role(app):
```
Do I need to say more? 😆

flowstef commented on flowauth/backend/tests/test_user_views.py:
```diff
@@ -33,69 +16,14 @@ def test_server_access(client, auth, test_user):
-    result = response.get_json()
-
-    assert [] == result
+# Replaece with test_no_roles
```
Why comment out?

flowstef commented on flowauth/frontend/src/Dashboard.jsx:
```diff
@@ -143,20 +145,19 @@ class Dashboard extends React.Component {
           <UserServer
-            serverID={this.state.activeServer}
-            serverName={this.state.activeServerName}
+            serverID = {this.state.activeServer}
```
Same as somewhere else: I'd keep the name and value in sync.

flowstef commented on flowauth/frontend/src/ServerAdminDetails.jsx:
```diff
@@ -116,14 +116,14 @@ class ServerAdminDetails extends React.Component {
-        } else {
-          this.setState({ hasError: true, error: err });
-        }
+        // if (err.code === 400) {
```
Uh-oh, swallowing exceptions is naughty!

flowstef commented on flowauth/frontend/src/ServerAdminDetails.jsx:
```diff
@@ -167,24 +168,27 @@ class ServerAdminDetails extends React.Component {
+        const rights = await capAwait.then(
+            resp => resp.reduce((p_x,x) => ({[x.name]:x.enabled, ...p_x}), {})
+          );
+        console.log("Rights", rights)
```
Same here: this looks like a tmp log; I'd either reduce it or remove it entirely.

flowstef commented on flowauth/frontend/src/ServerAdminDetails.jsx:
```diff
@@ -167,24 +168,27 @@ class ServerAdminDetails extends React.Component {
         const latestExpiry = (await limitsAwait).latest_token_expiry;
-        const maxLife = (await limitsAwait).longest_token_life;
+        const maxLife = (await limitsAwait).longest_token_life_minutes;
+        console.log(await limitsAwait)
```
That's curious - I haven't see await in console.log before.
Are you sure this is required?

Thingus replied:
I think if something's already been `await`'d, the result should just be returned as aif it was a normal variable - I guess strictly it doesn't need to be there, though

flowstef commented on flowauth/frontend/src/ServerAdminDetails.jsx:
```diff
@@ -167,24 +168,27 @@ class ServerAdminDetails extends React.Component {
           return {
             name: serverName,
-            rights: state.rights.length == 0 ? scopes : state.rights,
+            rights: state.rights.length == 0 
```
Better: `===0`

flowstef commented on flowauth/frontend/src/TokenDetails.jsx:
```diff
@@ -109,18 +110,18 @@ class TokenDetails extends React.Component {
-        permitted: (await rights).allowed_claims,
-        expiry: (await rights).latest_expiry,
-        latest_expiry: (await rights).latest_expiry,
+        rights: scopes,
```
I'd be tempted to rename the property too so it's the same as the value.

Thingus replied:
Two different properties, `rights` and `permitted`, that start from the same root `scopes`

flowstef commented on flowauth/frontend/src/TokenRolesPicker.jsx:
```diff
@@ -0,0 +1,67 @@
+        {roles.map((role, i) => (
+          <ListItem key={role.name}>
+            {role.name}
+          {/* <RoleString role={role}/> */}
```
Why comment this out?

flowstef commented on flowauth/frontend/src/UserAdminDetails.jsx:
```diff
@@ -48,14 +43,16 @@ class UserAdminDetails extends React.Component {
       if ((await json)["has_two_factor"]) {
         this.setState({ two_factor_can_be_disabled: true });
       }
+      //this.updateRoles(await json["roles"])
```
Why comment out?

flowstef commented on flowauth/frontend/src/UserAdminDetails.jsx:
```diff
@@ -48,14 +43,16 @@ class UserAdminDetails extends React.Component {
     } catch (err) {
       if (err.code !== 404) {
-        this.setState({ hasError: true, error: err });
+        this.setState({ hasError: true, errors: err });
```
If you expect more than one here should it not be an array and you add to it?
If it's just one why rename it?

greenape replied:
I think this is actually correcting it to match the state as declared at init: https://github.com/Flowminder/FlowKit/pull/5163/files#diff-91d75c71d0bfc65946adf1de0285ade03492ef5f84f3546a32b13761b329d56aL46

But, I think the correction should be the other way around - i.e. rename errors error

flowstef commented on flowauth/frontend/src/UserAdminDetails.jsx:
```diff
@@ -111,20 +112,19 @@ class UserAdminDetails extends React.Component {
-    this.setState({ servers: servers });
+
+  updateRoles = (roles) => {
+    console.log("Roles: ", roles)
```
Again I'd remove this.

flowstef commented on flowauth/frontend/src/UserAdminDetails.jsx:
```diff
@@ -143,32 +143,33 @@ class UserAdminDetails extends React.Component {
           this.setState({ pageError: true, errors: err });
         } else {
-          this.setState({ hasError: true, error: err });
+          this.setState({ hasError: true, errors: err });
```
Same as above - perhaps I'm misunderstanding it but it looks odd.

flowstef commented on flowauth/frontend/src/UserRolesPicker.jsx:
```diff
@@ -0,0 +1,77 @@
+
+
+  const handleChange = (event) => {
+    console.log("Selected roles: ", event.target.value)
```
It looks a bit verbose. Using console.debug (or whatever log level is relevant) would help to reduce clutter in the console.
Hoping that most of them would be disabled in production?

flowstef commented on flowauth/backend/flowauth/token_management.py:
```diff
@@ -80,23 +67,25 @@ def my_access(server_id):
-            for token in Token.query.filter(Token.owner == current_user)
-        ]
-    )
+#
```
Why comment this out?

flowstef commented on flowauth/backend/flowauth/token_management.py:
```diff
@@ -153,28 +142,57 @@ def add_token(server):
+        .all()
+        for role in json["roles"]
+    }
+    # claims = [scope.name for scope in Scope.query.filter(Scope.id.in_(scope_ids)).all()]
```
Just another complaint about commented out code 🙈

flowstef commented on flowauth/backend/flowauth/models.py:
```diff
@@ -22,32 +22,40 @@
 class User(db.Model):
     """
-    A user. Has at least one group.
+    A user.
```
This did make me chuckle 😄

flowstef commented on flowauth/backend/flowauth/models.py:
```diff
@@ -651,59 +575,60 @@ def make_demodata():
     test_server = Server(
         name="TEST_SERVER",
-        longest_token_life=2880,
+        longest_token_life_minutes=31 * 24 * 60,
```
I like this! Much more readable!

flowstef commented on flowauth/backend/flowauth/models.py:
```diff
@@ -651,59 +575,60 @@ def make_demodata():
-            latest_end=datetime.datetime.now() + datetime.timedelta(days=28),
+            scopes=[reader_scope, example_geo_scope],
+            latest_token_expiry=datetime.datetime.now() + datetime.timedelta(days=30),
+            longest_token_life_minutes=30 * 24 * 60,
```
Interesting you went with 30 here instead of 31 as above. Any reason for this?

Thingus replied:
Yep, obliviousness. I'm going to rework the demo data  over the next PR to bring it more into line with 'real' permissions, and I'll tidy up these when I do

flowstef commented on flowauth/backend/tests/conftest.py:
```diff
@@ -11,26 +12,47 @@
     "TestTwoFactorUser", TestUser._fields + ("otp_generator", "backup_codes")
 )
-TestGroup = namedtuple("TestGroup", ["id", "name"])
+
```
Uh-oh, more commented-out code...

flowstef commented on flowauth/backend/tests/conftest.py:
```diff
@@ -141,93 +159,124 @@ def test_two_factor_auth_user(app, get_two_factor_code):
         dummy_server_a = Server(
             name="DUMMY_SERVER_A",
-            longest_token_life=2,
+            longest_token_life_minutes=2880,
```
Would be nice to have it readable like in the other file.

flowstef commented on flowauth/backend/tests/test_role_admin.py:
```diff
@@ -0,0 +1,140 @@
+    }
+
+
+@pytest.mark.skip(
```
Why not delete the test?

Thingus replied:
Code review reasons: with mark.skip, reviewers can weigh in on whether getting rid of the test is reasonable or not.

flowstef commented on flowauth/frontend/src/RoleDetails.jsx:
```diff
@@ -0,0 +1,320 @@
+
+      const fetch_role = (async () => {
+        const role = await getRole(item_id);
+        console.log("Role fetched");
```
Same comment as in other file: log levels and verbosity

flowstef commented on flowauth/frontend/src/RoleDetails.jsx:
```diff
@@ -0,0 +1,320 @@
+      setNameHelperText("");
+      setNameIsValid(true)
+    } else if (name.length === 0) {
+      setNameHelperText("Role name can not be blank.");
```
I'd think it's cannot instead of can not.

flowstef commented on flowauth/frontend/src/RoleDetails.jsx:
```diff
@@ -0,0 +1,320 @@
+    if (maxLifetime.match(numbers)){
+      setLifetimeHelperText("");
+      setLifetimeIsValid(true)
+    } else if (maxLifetime == "") {
```
===

greenape replied:
```suggestion
    } else if (maxLifetime === "") {
```

flowstef commented on flowauth/frontend/src/RoleDetails.jsx:
```diff
@@ -0,0 +1,320 @@
+  }
+
+  const handleServerChange = (event) => {
+    console.log("Server picker event handled");
```
I think this may be ever so slightly too verbose.
OK during development but now I think it can safely go?!

flowstef commented on flowauth/frontend/src/RoleScopePicker.jsx:
```diff
@@ -0,0 +1,111 @@
+        c_role => s_scope.name.startsWith(c_role)
+      )
+    ))
+    // setRoleScopes(serverScopes.filter(x => (
```
<insert usual comment here>

flowstef replied:
`<insert usual comment here>`

flowstef commented on flowauth/frontend/src/RoleScopePicker.jsx:
```diff
@@ -0,0 +1,111 @@
+      options={rightsChoices}
+      value={selectedRights}
+      onChange={handleChange}
+      // label={"Scopes"}
```
Why remove the label?

flowstef commented on flowauth/frontend/src/TokenBuilder.jsx:
```diff
@@ -0,0 +1,178 @@
+      setNameHelperText("");
+      setNameIsValid(true)
+    } else if (name.length === 0) {
+      setNameHelperText("Role name can not be blank.");
```
Copy and paste :grimacing: 
That would be "Token name...", no?

flowstef commented on flowauth/frontend/src/util/api.js:
```diff
@@ -236,26 +217,39 @@ export async function editServer(
+    method: "POST",
+    body: JSON.stringify(new_scopes)
+  }
+  return await getResponse("/admin/servers/"+server_id+"/scopes", dat)
```
Why not use f-strings here?

flowstef replied:
(it's the same throughout the file)

flowstef commented on flowauth/frontend/src/util/util.js:
```diff
@@ -26,80 +27,144 @@ function zip() {
+  const things = Object.keys(tree).map((branch, index) => {
+    const this_label = label ==="" ? branch: [label, branch].join(".")
+    const this_value = this_label
+    // const this_value = value === "" ? index.toString(): [value, index.toString()].join("-")
```
Remove?

flowstef commented on flowauth/frontend/src/util/util.js:
```diff
@@ -26,80 +27,144 @@ function zip() {
+  const scopes_obj = {};
+  roles.forEach((role) => {
+    role.scopes.forEach((scope) => {
+      if (scopes_obj[scope] == undefined){
```
===

flowstef reviewed (commented):
I'm not as deep in this code as you (lucky) "backend" devs.
My feedback is mainly pointing out general stuff like removing commented out code or deep equality and some typos.
I did run it locally and it worked for me so that's good.

greenape commented:
Can we have a quick sequence diagram to show how the auth flow works now?

jc-harrison commented on flowapi/flowapi/permissions.py:
```diff
@@ -267,86 +154,98 @@ def schema_to_scopes(schema: dict) -> Iterable[str]:
+            query_kind: sub_query...}
+    }
+    returns the scope triplets of the query in the form "agg_unit:tl_query:sub_query".
+    Will always return "agg_unit:tl_query:tl_query"
```
I wonder whether it would be best to only return `agg_unit:tl_query:tl_query` if there are no sub-queries? (I.e. if a user has the required sub-query scopes, they don't need to also have the additional top-level scope). Effectively this would change the meaning of `agg_unit:tl_query:tl_query` from "can run a `tl_query` query" to "can run a `tl_query` query with no sub-queries".

Otherwise a role with scope `admin3:spatial_aggregate:daily_location` but not `admin3:spatial_aggregate:spatial_aggregate` would not be able to run any spatial aggregate query, but it's not necessarily clear to somebody creating a role in FlowAuth that they have to tick the `admin3:spatial_aggregate:spatial_aggregate` box otherwise any other selected `admin3:spatial_aggregate:X` scopes are meaningless.

An alternative could be to enforce this on the FlowAuth side (i.e. if `agg_unit:tl_query:sub_query` is selected, automatically add `agg_unit:tl_query:tl_query` to the role as well). But I think overall it's cleaner to handle this in FlowAPI.

jc-harrison commented on flowauth/frontend/src/util/util.js:
```diff
@@ -26,80 +27,144 @@ function zip() {
+
+function jsonify_inner(tree, label, value, enabled){
+  const things = Object.keys(tree).map((branch, index) => {
+    const this_label = label ==="" ? branch: [label, branch].join(".")
```
It's a minor point, but I'm not sure we need the label to be the full tree path at each level of the cascade. I.e. instead of
```
[] agg_unit -> [] agg_unit.tl_query -> [] agg_unit.tl_query.sub_query
```
I think it would be sufficient (and more concise) to display
```
[] agg_unit -> [] tl_query -> [] sub_query
```
(the layout of the MultiCascader itself shows the full path, so I don't think the labels need to).

jc-harrison commented on flowauth/frontend/src/RoleList.jsx:
```diff
@@ -14,5 +14,11 @@ const styles = (theme) => ({
 });
-const GroupList = Lister(GroupDetails, "Groups", getGroups, deleteGroup);
-export default withStyles(styles)(GroupList);
+const RoleList = Lister(
```
I think it would be good to display which server each role corresponds to (e.g. by having a heading per server with roles for that server listed underneath, or a server-name label alongside each role).

jc-harrison replied:
Overall I think I'd prefer to have a separate header per server, but if a different option is quicker to implement then that's fine too.

jc-harrison commented on flowauth/frontend/src/RoleDetails.jsx:
```diff
@@ -0,0 +1,320 @@
+          helperText={name_helper_text}
+        />
+      </Grid>
+      <MuiPickersUtilsProvider utils={DateFnsUtils}>
```
Expiry date defaults to now in CreateRole. It would be preferable for this to default to the server's "latest expiry" (and similarly for the max lifetime to default to the server's max lifetime).

jc-harrison replied:
I suppose the tricky part here is that the server is chosen within the "CreateRole" view, so isn't known when the view is first opened. In many ways I think it would be cleaner to have a separate "new role" button per server (within RoleList) and remove the server selection from within CreateRole. This could be combined with the "server headings in RoleList" suggestion from my earlier comment. But I appreciate that's not a trivial change to make.

Thingus replied:
`CreateRole` goes and fetches the role list for the appropriate server in current implementation anyway - so I think it should be fine to do.

Thingus replied:
Present in the new `roles` view

jc-harrison commented on flowauth/backend/flowauth/roles.py:
```diff
@@ -0,0 +1,162 @@
+        users=role_users,
+        server=server,
+        latest_token_expiry=json["latest_token_expiry"],
+        longest_token_life_minutes=json["longest_token_life_minutes"],
```
We need to enforce here that the latest expiry and longest lifetime for the role don't exceed those of the server. Previously we had frontend restrictions on this as well (i.e. dates after the server's latest expiry were greyed out in "Create group", and similar for max lifetime), but this at least needs to be enforced in the backend. Previous code that checked this for groups is here: https://github.com/Flowminder/FlowKit/blob/master/flowauth/backend/flowauth/groups.py#L214-L219.

Thingus replied:
In branch + test now

jc-harrison reviewed (commented):
A few issues I've encountered with create/edit role interactions:
- I can't create a role with no users (no error is raised, it just doesn't save the role).
  - Fixed in new Roles view - John
- I can't select just a subset of a tree of multi-part scopes (e.g. if I select "run", "admin3:spatial_aggregate:spatial_aggregate" and "admin3:spatial_aggregate:daily_location", and then save, only "run" will be assigned to the role. But if I select "admin3" (all), then these _will_ be assigned to the role).
- After saving a new role, the new role doesn't immediately appear in the roles list - I have to switch to another view and back for the new role to appear.
  - Fixed in new Roles view -John
- I can't remove a user from the pre-defined "viewer" and "runner" roles in the demo data, but I can do so for new roles that I define. I'm not sure why the behaviour is different for the pre-defined demo roles.

jc-harrison commented on flowauth/backend/flowauth/roles.py:
```diff
@@ -0,0 +1,162 @@
+    return jsonify(role_to_dict(role))
+
+
+@blueprint.route("/", methods=["POST"])
```
Now that roles are per-server (as opposed to groups, which could correspond to multiple servers), I wonder whether it would be more consistent to use `/roles/<server>` as the API route here (consistent with `/tokens/<server>`), instead of having the server ID in the json. This is not critical, though.

Thingus replied:
That might cause confusion with the GET and PATCH methods, which are on `/roles/<role_id>`.

jc-harrison replied:
Could change those to `/roles/<server_id>/<role_id>`, but I suppose that then requires remembering that it's `/roles/<server_id>/<role_id>` and not `/roles/<role_id>/<server_id>`. This change would make more sense if we were allowing role ID to be unique only per server (e.g. two different servers could each have an 'analyst' role), but if role ID alone is globally unique then having to specify both server_id and role_id in the route adds redundancy, so I think let's keep it as it is.

jc-harrison reviewed (commented):
When viewing the token list (or the "create token" view), there is nothing to show which server is selected. Previously the server name was displayed in a heading above the list of tokens - it would be good to add this back.
-In blanch now -John

jc-harrison commented on flowauth/backend/flowauth/roles.py:
```diff
@@ -0,0 +1,162 @@
+        elif key == "scopes":
+            value = [Scope.query.filter(Scope.id == sid).first() for sid in value]
+        elif key == "server":
+            value = Server.query.filter(Server.id == value).first()
```
I don't think this should allow editing the server for a role.  A role should be associated with a server on creation, and I don't think it would ever make sense to change this (not least because a different server may have a different set of scopes, so if you change the server for a previously-defined role then there would need to be a messy process of checking which of the role's scopes are still applicable for the new server).

Thingus commented:
`Add Token` save dialog permits blank token names

jc-harrison commented on flowapi/flowapi/user_model.py:
```diff
@@ -24,31 +23,35 @@ class UserObject:
-        for action in actions:
-            if {*scopes, action} in self.scopes:
-                return True
+    def has_access(self, *, actions=None, claims=None) -> bool:
```
Could combine `actions` and `claims` into a single arg here (e.g. `scopes`).

jc-harrison commented on flowauth/backend/tests/test_token_generation.py:
```diff
@@ -9,77 +9,114 @@
+
+        # Testing attempting reader on second server
+        response = client.post(
+            "/tokens/tokens/2", headers={"X-CSRF-Token": csrf_cookie}, json=token_eq
```
I think you want to request a token for `reader_role` here (i.e. a role that test_user does have access to, but for the wrong server). Currently `token_eq` is unchanged from the first part of the test.

jc-harrison commented on flowauth/backend/flowauth/token_management.py:
```diff
@@ -134,47 +102,76 @@ def add_token(server):
+    expiry = reduce(
+        lambda prev, cur: prev if prev < cur else cur,
+        (role.pop("latest_token_expiry") for role in roles),
+    )
```
This finds the earliest "latest expiry" across all applicable roles, which is good, but you also need to take into account the shortest "longest lifetime". (If I'm reading the models correctly, `Role.latest_token_expiry` directly returns the "latest token expiry" setting and is independent of the "longest_token_life_minutes", unlike `User.latest_token_expiry(server)` which calculates latest expiry from a combination of "latest expiry", "longest lifetime" and `datetime.now()`).

Thingus commented:
New diagram in PR text

flowstef reviewed (approved):
Looks good to me.

flowstef reviewed (commented):
Not sure what to look for here - the regex seems to work even though the check fails.
Maybe it could be more elegant, don't know, but if it works I think it's fine.

jc-harrison commented on flowapi/flowapi/permissions.py:
```diff
@@ -267,86 +160,103 @@ def schema_to_scopes(schema: dict) -> Iterable[str]:
+    if reply["status"] != "success":
+        raise JSONHTTPException(description=reply["msg"])
+    try:
+        if reply["payload"]["aggregation_unit"] == "None":
```
```suggestion
        if reply["payload"]["aggregation_unit"] is None:
```

flowstef commented on flowauth/backend/flowauth/main.py:
```diff
@@ -85,20 +85,18 @@ def handle_invalid_usage(error):
-
     print(error)
-    _, _, error_message = error.args[0].partition(" ")
+    _, _, error_message = error.args[-1].partition(" ")
```
That line looks a bit cryptic to me and the `[-1]` to me is asking for an exception but I'll say I trust you know what you're doing.

Thingus replied:
Point - sorted.

flowstef reviewed (commented):
See individual comments.

jc-harrison commented on docs/source/developer/adr/0012-claims-role-scope-rework.md:
```diff
@@ -0,0 +1,72 @@
+
+**Status**
+
+Proposed
```
I think perhaps "Accepted" would be a more appropriate status.

jc-harrison commented on docs/source/developer/adr/0012-claims-role-scope-rework.md:
```diff
@@ -0,0 +1,72 @@
+ - *Simple scopes*
+
+   These control access to a server-wide capability and are all-or-nothing. The currently implemented simple scopes are:
+   - `run`; permits a user that olds this role to run queries using the `run` flowAPI endpoint
```
```suggestion
   - `run`; permits a user that holds this role to run queries using the `run` flowAPI endpoint
```

jc-harrison commented on docs/source/developer/adr/0012-claims-role-scope-rework.md:
```diff
@@ -0,0 +1,72 @@
+ - *Complex scopes*
+  
+   These control access to a combination of a geographic component, a top-level query and a descendent query.
+   - Top-level queries are the methods that are avilable to the users of Flowmachine directly.
```
```suggestion
   - Top-level queries are the methods that are available to the users of Flowmachine directly.
```

jc-harrison commented on docs/source/developer/adr/0012-claims-role-scope-rework.md:
```diff
@@ -0,0 +1,72 @@
+   These control access to a combination of a geographic component, a top-level query and a descendent query.
+   - Top-level queries are the methods that are avilable to the users of Flowmachine directly.
+   - Each top-level query may require a set of sub-queries to run, wich in turn may require sub-queries of their own; these are the descendent queries.
+   - Finally, the geographic component is the spatial presentation that is available to the role.
```
Maybe worth mentioning that we use "nonspatial" for scopes relating to queries without spatial aggregation.

jc-harrison commented on docs/source/developer/adr/0012-claims-role-scope-rework.md:
```diff
@@ -0,0 +1,72 @@
+   - Each top-level query may require a set of sub-queries to run, wich in turn may require sub-queries of their own; these are the descendent queries.
+   - Finally, the geographic component is the spatial presentation that is available to the role.
+   
+   Complex scopes are of the form `geographic_component:top_level_query:sub_query` - this ordering has been built on assumed order of importance to users. For example it will be a more common use case that an administrator will want to create a role that restrics access only to admin levels 0 or 1 than to the `most_frequest_location` sub-query of the `spatial_aggregate` TL query.
```
```suggestion
   Complex scopes are of the form `geographic_component:top_level_query:sub_query` - this ordering has been built on assumed order of importance to users. For example it will be a more common use case that an administrator will want to create a role that restrics access only to admin levels 0 or 1 than to the `most_frequest_location` sub-query of the `spatial_aggregate` top-level query.
```

jc-harrison commented on docs/source/developer/adr/0012-claims-role-scope-rework.md:
```diff
@@ -0,0 +1,72 @@
+**Consequences**
+ - Much faster and more lightweight building of tokens
+ - Flowauth frontend can now be used on lower-capability machines
+ - Loss of fine-grained control over query permissions
```
```suggestion
 - Less fine-grained control over query permissions
```

jc-harrison replied:
I also think it's worth mentioning that there's a positive side to this one as well as a negative, because reducing the number of possible scopes makes decisions about what scopes should be granted to a role more manageable - the auth system consequently becomes more user-friendly (so the granularity that remains is hopefully _more_ likely to be utilised, rather than just granting all permissions because there are too many to select from).

jc-harrison commented on docs/source/developer/adr/0012-claims-role-scope-rework.md:
```diff
@@ -0,0 +1,72 @@
+ - Much faster and more lightweight building of tokens
+ - Flowauth frontend can now be used on lower-capability machines
+ - Loss of fine-grained control over query permissions
+ - Users can't be grouped anymore (roles supercede group functions)
```
I think roles do still "group" users in some sense. I'd say the main change in this regard is that permission scopes can no longer be assigned directly to users (only via roles).

jc-harrison reviewed (commented):
A few comments on the ADR here

greenape commented on flowauth/backend/flowauth/token_management.py:
```diff
@@ -134,47 +102,65 @@ def add_token(server):
+    # feature todo: flag this to the user
+    # This isn't about the user, so get these values from the server
+
+    current_app.logger.debug("token_expiry")
```
```suggestion
```

greenape commented on flowauth/backend/tests/conftest.py:
```diff
@@ -141,93 +134,132 @@ def test_two_factor_auth_user(app, get_two_factor_code):
+    with app.app_context():
+        dummy_server_a, dummy_server_b = test_servers
+        scopes = [
+            read_scope_a := Scope(
```
Any reason we need the walrus operator here?

greenape commented on flowauth/backend/tests/conftest.py:
```diff
@@ -141,93 +134,132 @@ def test_two_factor_auth_user(app, get_two_factor_code):
+        return runner, reader, reader_b
+
+
+@pytest.fixture  # (scope="session")
```
```suggestion
@pytest.fixture
```

There are some other instances of commenting out in preference to deleting - better to delete, git remembers it was there.

greenape commented on flowauth/backend/tests/test_auth.py:
```diff
@@ -6,25 +6,26 @@
-        assert g.user.username == username
+def test_login(app, client, auth, test_user):
+    with app.app_context():
+        """Test that we can log in by posting a username and password as json"""
```
This should be up a line I think

greenape commented on flowauth/frontend/src/RoleScopePicker.cy.jsx:
```diff
@@ -0,0 +1,139 @@
+/* eslint-disable quotes */
+/* eslint-disable no-undef */
+/* eslint-disable react/react-in-jsx-scope */
+import RoleScopePicker from "./RoleScopePicker";
```
Does this get run on CI btw?

greenape reviewed (changes requested):
Still some cleanup to do here I think, and can you make sure you've checked all extant comments as well pls.

Thingus commented on flowauth/backend/flowauth/main.py:
```diff
@@ -85,20 +85,18 @@ def handle_invalid_usage(error):
         name = error.params[0]
         out_msg = f"{table.capitalize()} '{name}' already exists on server"
-        return dict(status=400, statusText=out_msg), 400
+        return dict(status=399, statusText=out_msg), 400
```
Nope - now returns 400 in both

Thingus commented on flowapi/flowapi/query_endpoints.py:
```diff
@@ -67,7 +67,17 @@ async def run_query():
+    try:
+        await current_user.can_run(query_json=json_data)
+    except KeyError as e:
+        return (
```
It gets raised in `query_to_scopes` (`permissions.py:250`)

greenape replied:
In which case, presumably that could/should be raising it as a badquery? I mean the function knows it is meant to be dealing with a query, the query object is (presumably) malformed.

Thingus replied:
I think I'm missing something here, so I'll clarify; `query_to_scopes` runs the following:
```python
    try:
        if "query_kind" not in query_dict.keys():
            raise MissingQueryKindError
        tl_query_name = query_dict["query_kind"]
        query_list = grab_on_key_list(query_dict, ["query_kind"])
    except Exception:
        raise BadQueryError
```
`query_to_scopes` is called by `can_run` (and the rest of the `can_` functions)
, raising a BadQueryError in this case. I guess that this means that we could remove the `KeyError` catch.

Thingus commented on flowauth/backend/flowauth/models.py:
```diff
@@ -387,7 +376,105 @@ def backup_code(self, plaintext: str):
+    name = db.Column(db.String)
+    enabled = db.Column(db.Boolean, default=True)
+    server_id = db.Column(db.Integer, db.ForeignKey("server.id"))
+    # TODO: Make sure scopes are unique within server
```
No; fixed by `__table_args__ ...`

Thingus commented on flowauth/backend/flowauth/servers.py:
```diff
@@ -108,9 +174,11 @@ def list_server_time_limits(server_id):
     on a server.
     """
     server = Server.query.filter(Server.id == server_id).first_or_404()
+    # I have no idea why this test is failing, it's been like this since before
```
Nope (endpoint called in `test_server_admin.test_get_server_time_limits`)

greenape commented on flowapi/flowapi/permissions.py:
```suggestion
def _search_for_nested_keys(in_iter, search_keys):
```

greenape commented on flowauth/backend/flowauth/permissions.py:
```diff
@@ -0,0 +1,248 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
```
Is this file actually used?

greenape commented on flowauth/backend/flowauth/roles.py:
```diff
@@ -0,0 +1,185 @@
+                raise InvalidUsage(
+                    "Role cannot have a maximum lifetime greater than server"
+                )
+        elif key not in ["name", "server_id"]:
```
Does the ability to change the server id mean that you can create a role with a later expiry than should be allowed?

greenape replied:
And presumably invalid scopes, now I think about it.

Thingus replied:
As in, on server A (exp. Now+10) you create role 1 (exp. Now+10) and then move it to server B (exp. Now+1)?
Yes. I think that being able to move roles between servers is also more trouble than it's worth, so 'server_id' will come out.

Thingus commented on flowauth/backend/flowauth/servers.py:
```diff
@@ -35,70 +39,132 @@ def get_server(server_id):
 @login_required
 @admin_permission.require(http_exception=401)
-def edit_server_capabilities(server_id):
+def get_role(server_id, role_id):
```
Nope - removed to no apparent ill effects

Thingus commented on flowauth/backend/flowauth/servers.py:
```diff
@@ -35,70 +39,132 @@ def get_server(server_id):
+    server_scopes = Scope.query.filter(Scope.server_id == server_id).all()
+    for scope in server_scopes:
+        db.session.delete(scope)
+    db.session.commit()
```
Yep, good catch.

Thingus replied:
Turns out we do; if we don't commit the delete first, `session.add` gets caught by the `UNIQUE` constraint in `models.Scope`.

greenape commented on flowapi/flowapi/permissions.py:
```diff
@@ -1,259 +1,72 @@
-    for walk in tree_walk:
-        yield from tree_walk_to_scope_list(walk)
-
+def grab_on_key_list(in_iter, search_keys):
```
I think this func could really do with a doctoring, because it is potentially pretty difficult to follow

greenape commented on flowauth/frontend/cypress/fixtures/openapi.json:
```diff
@@ -0,0 +1,3173 @@
+{
+  "components": {
```
Should this file be here? It seems to be a broken spec, and better to use the api_sec.json which symlinks to the one in the integration tests that gets checked against the one generated by flowapi.

greenape commented on flowclient/flowclient/async_connection.py:
```diff
@@ -86,9 +86,7 @@ def update_token(self, token: str) -> None:
-            self.user = jwt.decode(token, options=dict(verify_signature=False))[
-                "identity"
-            ]
+            self.user = jwt.decode(token, options=dict(verify_signature=False))["sub"]
```
I reckon this warrants a changed entry in the changelog.

greenape commented on flowapi/flowapi/config.py:
```diff
@@ -73,4 +73,5 @@ def get_config():
         FLOWMACHINE_PORT=flowmachine_port,
         FLOWDB_DSN=f"postgresql://{flowdb_user}:{flowdb_password}@{flowdb_host}:{flowdb_port}/flowdb",
         JWT_DECODE_AUDIENCE=flowapi_server_id,
+        JWT_IDENTITY_CLAIM="sub",
```
Also warrants a changed entry.

greenape commented on flowauth/backend/flowauth/token_management.py:
```diff
@@ -134,47 +102,63 @@ def add_token(server):
+        username=current_user.username,
         private_key=current_app.config["PRIVATE_JWT_SIGNING_KEY"],
+        lifetime=token_expiry - datetime.datetime.now(),
+        roles={role.name: sorted([ss.name for ss in role.scopes]) for role in roles},
```
```suggestion
        roles={role.name: sorted(ss.name for ss in role.scopes) for role in roles},
```

Don't need the comprehension here really.

Thingus commented on flowauth/backend/flowauth/models.py:
```diff
@@ -645,65 +606,74 @@ def make_demodata():
+            "x-security-scopes"
+        ]
+    ]
+    reader_scope = list(filter(lambda scope: scope.name == "get_result", scopes))[0]
```
Yeah, `scopes` is now a dict keyed on scope.name. Much cleaner.
